ARROW-14744: [R] open_dataset() error when `schema` argument supplied, but `column_names` not supplied to `CSVReadOptions`

toppyy opened the pull request:
Supplement CsvReadOptions with column_names from schema if column_names are not set in read_options

thisisnic commented:
Just pasting here the conversation from JIRA:

> I did, however, run into trouble. Say, for example, the user has set skip_rows-option like this:
`read_options=arrow::CsvReadOptions$create(skip_rows=1))`
I image we'd like to keep whatever options the user has set when we re-create the `CsvReadOptions` object  with column names from the schema. The problem is that I cannot access `skip_rows` in the object after it's created, so I cannot use that information to create another instance of `CsvReadOptions` that has both the `column_names` and `skip_rows` set (plus any other options).

> Any thoughts? Is there a way to access `skip_rows` and other attributes that I'm unaware of? Of course, one solution is to change class declaration of `CsvReadOptions` to have access to these attributes.

Thanks for opening this draft PR!  After checking out a copy of your branch, I understand what's going on here and why it's not working a lot better not.  

Currently, the approaches that come to mind for me are:
1. as you say, update that class so we can modify contents after creation
2. update the signature of `CsvFileFormat$create()`, setting `read_options` to a default value of `NULL` and then calling `csv_file_format_read_opts()` later in the function, passing in both the schema and any user-set values.

I haven't fully fleshed out the second option or tested to see if it'll work, but if it does I'd be in favour of doing it that way so we can make the change we need without having to modify the structure of the existing classes.

toppyy commented:
Thanks for copy-pasting the discussion - it didn't occur to me. I tried option 2, but ended up having to change the class CsvReadOptions also (see below why). So the latest commits really implement options one and two. 

To me it still seems that it's impossible to get the value of _skip_rows_ later on if user has set it like this:
> `open_dataset(
>   td,
>   format = 'csv',
>   read_options = arrow::CsvReadOptions$create(
>     skip_rows = 1
>   )
> )`

Because of this I had to implement CsvReadOptions$skip_rows to be able to retrieve the value in `csv_file_format_read_opts`. I could be wrong though.

The functionality is as follows:
- Check if column_names are given as a separate argument. If so, use them.
- Check if column_names are given in read_options. If so, use them.
- Check if schema is set. If so, use names from schema as column names. 

For skip rows, we do the same (with the exception of using schema).

The code is a bit verbose and may be improved by some simplification. Also, if any other options (in addition to column_names & skip_rows) are introduced, it needs an update - so not ideal in that sense.

Let me know what you think!

thisisnic commented:
Thanks for the updates here @toppyy .  I've taken the time to have a proper think about this, and on reflection, I don't think we need to make `open_dataset( td, format = 'csv', read_options = CsvReadOptions$create( skip_rows = 1 ))` work for users as they can pass in their `skip_rows` parameter this way: `open_dataset( td, format = 'csv', skip_rows = 1)`.

Directly using `CsvReadOptions$create()` is pretty low-level and so here we can probably assume that someone using it is responsible for making sure things match up themselves (though should absolutely add further documentation to show our recommended way that users work with `open_dataset()` so it's clear - perhaps that could be part of this PR if you're interested but no worries if not).

I feel an alternative solution here might be just to check if there are conflicting arguments specified to `open_dataset()` (e.g. through specifying the `read_options` argument in the ellipses as well as individual read options).  It might be something along the lines of adding validation at the end of open_dataset so that if `(!is.null(schema))` and format is csv, ensure that `identical(names(schema), read_options$column_names)` or raise an error.

Does that make sense?

toppyy commented:
I think that makes a lot sense! Trying to infer the column names from schema added (a surprising) amount of complexity to the code for little to none added value for the user. As you say, they can pass the arguments using a different approach. 

Just to make sure I got it right before I jump into making changes, the approach we're taking is this:
Instead of deriving the column_names from the schema, we raise an error if `CsvReadOptions$create()` is used for read_options and it has no column names or they differ from the schema?

If so, could the check for column names be inside `CsvFileFormat$create`? As the issue is limited to formats that use this function (not relevant for parquet).

I can write this  up in the documentation as a part of this PR.

thisisnic commented:
Yeah, spot on, and there is an extra level of complexity added here due to the fact that the way in which we infer column names from a schema is a bit incorrect (but this is being dealt with as part of another issue)

Are you happy with leaving the existing code which derives column names from a schema as it was before any of the changes in this PR, and as you say, raise an error if CsvReadOptions$create() is used for read_options but is not consistent with the schema?

There's a PR which was just merged that does a similar thing but relates to partitioning, and is a great example of this kind of thing being done really nicely: https://github.com/apache/arrow/blob/99f7c3cf3e6c2a9555ceff3d48ef73e485ede546/r/R/dataset-factory.R#L85-L95

What you said about where the code for this should go sounds right and thanks for volunteering to update the docs too!

Thanks for sticking with this even though we've drastically changed what we're doing to resolve this - even though we're not using the code from your previous solutions, the process of testing them out and the surrounding discussion has been really helpful for identifying some serious enhancements that can be made to how the schema and column name components interact and also how we direct our users to work with `open_dataset()`.

toppyy commented:
I reverted the previous commits and implemented the following logic in `CsvFileFormat$create` as discussed: If a schema is specified and column names are set in read_options, we raise an error if the two do not match. 

Also added a note of this in the documentation. 

I thought about the documentation aspect a bit. As a starting point, I added link to [read_csv_arrow](https://arrow.apache.org/docs/r/reference/read_delim_arrow.html) into `open_dataset` docs. It has some explanation on how to specify column names and types. Though I suspect this might be a bit misleading for the readers who are unaware of the fact that both functions end up calling `CsvReadOptions$create`.

It seems to that the `open_dataset` docs (or the cookbook) might benefit from additional examples or a separate section (like in read_csv_arrow) on this issue?

thisisnic commented on r/R/dataset-format.R:
```diff
@@ -122,6 +122,18 @@ CsvFileFormat$create <- function(...,
+  schema  <- options[["schema"]]
+
+  if (length(read_options$column_names) > 0 & !is.null(schema) & !identical(names(schema), read_options$column_names)) {
+    abort(c(
+        '"column_names" in read_options do not match the schema.',
+      i = "Set column_names in read_options to match the schema",
+      i = "Omit the read_options argument"
+    ))
```
Looking good, just a couple of suggestions:
- the user of the `open_dataset()` function won't necessarily have used the function `read_options()` directly, so we should remove the reference to it to avoid confusion
- to provide maximal useful information to the end-user, could you print out the arguments which are mismatches between `column_names` and the schema?

thisisnic replied:
Also, please could you add a test which triggers this error?

toppyy replied:
Thanks for the good comments!

It raises an error and prints out the mismatches between read_options$column_names and names(schema).

At the moment a mismatch is either due to 1) set difference or a 2) different order of column names. However, I'm a bit unsure of the latter. Do we in fact want to raise an error if the order of names differ? For example: `c("a","b") != c("a","b")`

Also, if this is the case, I'd be interested if you have any tips on convenient ways to do this. I resorted to using `suppressWarnings` to handle cases where vector lengths differ.

thisisnic commented on r/R/dataset.R:
```diff
@@ -123,6 +123,7 @@
 #' or call [`$NewScan()`][Scanner] to construct a query directly.
 #' @export
 #' @seealso `vignette("dataset", package = "arrow")`
+#' See [read_csv_arrow()] on how to specify column names and types for "csv"/"text" and "tsv" -formats.
```
This is great, though I think it might fit better after the content on line 121 along with links to `read_feather()` and `read_parquet()` and a more generic comment about viewing those page for format-specific options.

thisisnic reviewed (commented):
All looks good, just a couple of minor suggestions!  Thanks for this.

thisisnic commented:
> It seems to that the `open_dataset` docs (or the cookbook) might benefit from additional examples or a separate section (like in read_csv_arrow) on this issue?

Agreed - you're welcome to submit extra examples with the docs as part of this PR or we can open up a separate issue for that? Let me know which you'd rather do.

The cookbook is in a separate [repo](https://github.com/apache/arrow-cookbook), so I've opened a ticket there: https://github.com/apache/arrow-cookbook/issues/120

At present there is no chapter on datasets - I was thinking of starting one soon, but if you'd like to get it started, or contribute some content once it's started, let me know on that ticket.

toppyy commented:
> you're welcome to submit extra examples with the docs as part of this PR or we can open up a separate issue for that?

I think I'd prefer a separate issue just for the sake of clarity on my end. 

Thanks for the issue for the cookbook. I'll get back to that later after this one is sorted :)

toppyy commented:
I refactored/simplified the column_names vs. schema-names comparison a bit.

While doing this, I realized that the solution did not solve the original issue. I was comparing column names versus schema only when both are set. However, the original issue presented itself when read_options was set _without_ column_names. Like so:
`open_dataset(
  td,
  format='csv',
  schema = diamond_schema,
  skip_rows = 1,
  read_options=arrow::CsvReadOptions$create(
    skip_rows = 1
    # ..and no column_names
  )
) %>%  collect()`

To raise an error in situation also, the code now checks that `identical(read_options$column_names,names(schema))` in all scenarios where schema is set. So the example would create an error despite the fact that `column_names` is not set at all. I added a hint for this situation in the error message ("Omit the read_options -argument").

thisisnic commented on r/R/dataset-format.R:
```diff
@@ -122,6 +122,31 @@ CsvFileFormat$create <- function(...,
+
+    abort(c(
+            paste(
+                "column_names not matching or not found in schema-names:",
```
What do you think to the idea of rephrasing this to use "should" or "must" as per the tidyverse style guide (https://style.tidyverse.org/error-messages.html)?

thisisnic commented on r/R/dataset-format.R:
```diff
@@ -122,6 +122,31 @@ CsvFileFormat$create <- function(...,
+                "column_names not matching or not found in schema-names:",
+                deparse1(c(mismatch_colnames, not_in_schema))
+            ),
+            i = "Set column_names to match names of schema",
+            i = "Omit the column_names argument",
+            i = "Omit the read_options argument"
```
I really love that you've come up with helpful solutions for the end-user here.  However, I have a concern that given that there are multiple paths that users can take to get here, and `read_options` isn't an argument that the user is necessarily aware of (it's a few levels down in a function they wouldn't be likely to call directly), and so this error message might be a little confusing - it may be sufficient just to state what's wrong, without any suggestions.

Could I get your thoughts here @jonkeane?

jonkeane replied:
I agree with @thisisnic, I really like solutions provided in error messages generally. But here, since there are so many ways one can get to this (and not all of the solutions would work in all of those paths), it's probably better to simply error with the problem up above

thisisnic commented on r/tests/testthat/test-dataset-csv.R:
```diff
@@ -299,3 +299,21 @@ test_that("open_dataset() deals with BOMs (byte-order-marks) correctly", {
+  opts <- CsvReadOptions$create(column_names = c("i", "d"))
+  schema             <- schema(int = int32(), dbl = float64())
+
+  expect_error(
+    open_dataset(csv_dir, format = "csv", schema = schema, read_options = opts),
+    "column_names not matching or not found in schema-names"
+  )
```
Thanks for the test, looking good.  One minor suggestion: how about swapping this for a simpler case, e.g.:
```
open_dataset(csv_dir, format = "csv", schema = schema, column_names = c("these", "are", "not", "the", "columns", "from", "the", "schema"))
```
It then removes the need to do the more complex work of `opts <- CsvReadOptions$create(column_names = c("i", "d"))` which is technically off-label usage that we don't recommend anyway.

toppyy replied:
Thanks again for all of the comments, really helpful!

At the moment the example you provided would *not* raise an error if I've got it right. This is due to the following logic in `csv_file_format_read_opts`:

https://github.com/apache/arrow/blob/0fa4b9ca1e7b13ac230f85075d86d783eeea6a74/r/R/dataset-format.R#L223-L225

If schema is set, it overrides `column_names` in all situations. Because the check is done after `csv_file_format_read_opts` no error is raised despite of conflicting arguments from the user (which might be a little confusing for the user?).

thisisnic replied:
Good catch - let's update that to something like `if (!is.null(schema)) && is.null(opts[["column_names"]])`.

thisisnic reviewed (commented):
A few suggestions, but overall looking in good shape!

There is currently a CI failure for this PR - it's caused by the fact that we have a test for an error message that is triggered by `csv_file_format_parse_options()`:

https://github.com/apache/arrow/blob/daa5c18e9697a6455a7a75fec19594543c17b21e/r/tests/testthat/test-dataset-csv.R#L216-L219

This normally works fine as `csv_file_format_parse_options()` is called before `csv_file_format_read_opts()` but now it's being done in a different order, we don't get the desired error message.

If you call `devtools::test()` after your edits (or `make test` from the R directory), you can run these tests locally.

My suggestion here would be to use `force()` ahead of your new code, so you can determine the order in which those arguments are evaluated - if this is unclear, let me know and I'll explain more about what's going on.

thisisnic commented on r/R/dataset-format.R:
```diff
@@ -122,6 +122,30 @@ CsvFileFormat$create <- function(...,
+    not_in_schema <- setdiff(schema_names, column_names)
+
+    abort(
+      paste(
+        "column_names must match schema-names:",
+        deparse1(c(mismatch_colnames, not_in_schema))
+      )
```
```suggestion
      paste(
        "Values in `column_names` must match schema field names"
      )
```

thisisnic replied:
I've just removed the pasting of the fields as it's a little unclear which are missing from which, so best to leave out to reduce ambiguity!

thisisnic commented on r/tests/testthat/test-dataset-csv.R:
```diff
@@ -299,3 +299,20 @@ test_that("open_dataset() deals with BOMs (byte-order-marks) correctly", {
+
+  expect_error(
+    open_dataset(csv_dir, format = "csv", schema = schema, column_names = c("these", "wont", "match")),
+    "column_names must match schema-names"
```
```suggestion
    "Values in `column_names` must match schema field names"
```
Updating this to match change made above

thisisnic commented on r/R/dataset-format.R:
```diff
@@ -122,6 +122,30 @@ CsvFileFormat$create <- function(...,
+
+  if (!is.null(schema) & !identical(schema_names, column_names)) {
+
+    # Element wise comparison and set differnce of column_names and names in schema
+    mismatch_colnames <- column_names[match(column_names, schema_names, nomatch = 0) != seq(column_names)]
+    not_in_schema <- setdiff(schema_names, column_names)
```
```suggestion
    
```
Removing these as no longer needed due to changes below

thisisnic reviewed (approved):
I've made a few tweaks, but otherwise, this is looking great.  I'll merge once the CI passes.  Thanks for getting involved @toppyy and sticking with this!

thisisnic commented on r/R/dataset-format.R:
```diff
@@ -122,6 +122,27 @@ CsvFileFormat$create <- function(...,
+  schema_names <- names(schema)
+
+  if (!is.null(schema) & !identical(schema_names, column_names)) {
+
+    
```
```suggestion
```
